src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -311,21 +311,42 @@ public async Task RefreshAsync(CancellationToken cancellationToken)
                         applicationSettings = new Dictionary<string, ConfigurationSetting>(_applicationSettings, StringComparer.OrdinalIgnoreCase);
                         watchedSettings = new Dictionary<KeyValueIdentifier, ConfigurationSetting>(_watchedSettings);
 
-                        foreach (KeyValueChange change in keyValueChanges.Values.Concat(changedKeyValuesCollection))
+                        foreach (KeyValueWatcher changeWatcher in cacheExpiredWatchers.Concat(cacheExpiredMultiKeyWatchers))
                         {
-                            if (change.ChangeType == KeyValueChangeType.Deleted)
+                            UpdateCacheExpirationTime(changeWatcher);
+                        }
+
+                        foreach (KeyValuePair<KeyValueIdentifier, KeyValueChange> kvp in keyValueChanges)
+                        {
+                            KeyValueChange keyValueChange = kvp.Value;
+
+                            if (keyValueChange.ChangeType == KeyValueChangeType.Modified)
                             {
-                                applicationSettings.Remove(change.Key);
+                                applicationSettings[keyValueChange.Key] = keyValueChange.Current;
+                                watchedSettings[kvp.Key] = keyValueChange.Current;
                             }
-                            else if (change.ChangeType == KeyValueChangeType.Modified)
+                            else if (keyValueChange.ChangeType == KeyValueChangeType.Deleted)
                             {
-                                applicationSettings[change.Key] = change.Current;
+                                applicationSettings.Remove(keyValueChange.Key);
+                                watchedSettings.Remove(kvp.Key);
                             }
 
                             // Invalidate the cached Key Vault secret (if any) for this ConfigurationSetting
                             foreach (IKeyValueAdapter adapter in _options.Adapters)
                             {
-                                adapter.InvalidateCache(change.Current);
+                                adapter.InvalidateCache(keyValueChange.Current);

```

Comment:
Historically, we have invalidated adapter cache whenever key-values or feature flags change. But I dont think we need to invalidate adapter cache for feature flag changes because we dont cache feature flags in Key Vault adapter, and all other adapters dont have any caching at all.
That being said, I dont see an issue with invalidating the adapter cache for feature flag changes - it would have no effect on anything, and it would prevent any bugs in future if we ever decide to implement `InvalidateCache` for feature management adapter. This behavior would also be consistent with what we have in the main and preview branches currently. 
Can you add the same cache invalidation logic to the `changedKeyValuesCollection` loop below?